controllers/machine/machine_controller.go

```diff
@@ -0,0 +1,231 @@
+	ClusterId         string
+}
+
+// Reconcile reads that state of the cluster for machine objects and makes changes based the contained data
+func (r *MachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", req.Namespace, "Request.Name", req.Name)
+	reqLogger.Info("Reconciling Machine")
+
+	// Fetch the machines in openshift-machine-api
+	machine := &machinev1beta1.Machine{}
+	err := r.Client.Get(ctx, client.ObjectKey{Namespace: machineNamespace, Name: req.Name}, machine)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Info("Machine not found. Ensuring that no metrics for this machine are leftover")
+			// use the req.Name here because the Machine does not exist and will be nil
+			r.MetricsAggregator.RemoveMachineMetrics(req.Name)
+			return utils.DoNotRequeue()
+		}
+		reqLogger.Error(err, "An error occurred getting the machine")
+		return utils.RequeueWithError(err)
+	}
+
+	if machine != nil && machine.Status.Phase != nil && *machine.Status.Phase == "Deleting" {
+		reqLogger.Info("Found machine in deleting state. Looking for customer pods failing to delete")
+		ctx = context.WithValue(ctx, "machine", machine)
+		ctx = context.WithValue(ctx, "logger", reqLogger)
+		return r.evaluateDeletingMachine(ctx)
+	}
+
+	reqLogger.Info("Reconcile Complete")
+	return utils.DoNotRequeue()
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *MachineReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	indexerFunc := func(rawObj client.Object) []string {
+		event := rawObj.(*corev1.Event)
+		return []string{event.InvolvedObject.Name}
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Event{}, "involvedObject.name", indexerFunc); err != nil {
+		return err
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&machinev1beta1.Machine{}).
+		Complete(r)
+}
+
+func getMachineFromCtx(ctx context.Context) (*machinev1beta1.Machine, error) {
+	machineRaw := ctx.Value("machine")
+	if machineRaw == nil {
+		return nil, fmt.Errorf("Machine does not exist in context")
+	}
+
+	machine, ok := machineRaw.(*machinev1beta1.Machine)
+	if !ok {
+		return nil, fmt.Errorf("Could not cast machine context value to Machine type")
+	}
+
+	return machine, nil
+}
+
+func getLoggerFromCtx(ctx context.Context) (logr.Logger, error) {
+	loggerRaw := ctx.Value("logger")
+	if loggerRaw == nil {
+		return logr.Logger{}, fmt.Errorf("Logger does not exist in context")
+	}
+
+	logger, ok := loggerRaw.(logr.Logger)
+	if !ok {
+		return logr.Logger{}, fmt.Errorf("Could not cast logger context value to Logger type")
+	}
+
+	return logger, nil
+}
+
+func getMostRecentDrainFailedEvent(reqLogger logr.Logger, eventList *corev1.EventList) *corev1.Event {
+	// if there are no events at all exit now
+	if len(eventList.Items) == 0 {
+		reqLogger.Info("No events")
+		return nil
+	}
+
+	var newestEvent *corev1.Event
+
```

Comment:
Sure - that was my first thought as well - I just couldn't actually find any prior art on how to do that within go, granted I didn't look _too_ hard after my first initial searches came up empty.  

Because of that failure to find it, I actually was able to use a Test-Driven Development approach with this function, and was able to define the tests and write the function at the same time which was actually a nice little experience because I haven't done that in a long time haha :)

Reply:
This is pseudocode written directly in this comment and not tested, but you should be able to do something like...

``` go
sort.Slice(eventList.Items, func(i, j event) bool {
  return eventList.Items[i].LastTimeStamp < eventList.Items[j].LastTimeStamp
})
```

Reply:
I'll leave this up to y'all to tell me if this looks any simpler - I'm not seeing it as any easier to read but I'm also biased because I wrote it XD

```
	sort.Slice(eventList.Items, func(i, j int) bool {
		return eventList.Items[i].LastTimestamp.Time.After(eventList.Items[j].LastTimestamp.Time)
	})

	for i := range eventList.Items {
		event := &eventList.Items[i]
		if event.Reason == "DrainRequeued" {
			if strings.Contains(event.Message, "error when evicting pods") {
				newestEvent = event
				break
			}
		}
	}
```

Reply:
I'd say leave the loop as is. I personally find this approach a bit easier to understand. But now that we're on the topic of readability:
I would like to see the conditions of the if-statments being factored out into a separate `IsErrorEvictingPodsEvent(...)`-function, so we can have a unit test for this detection methodology specifically, cause i find this to be one of the core features.

Also, I find this particular example easier to read when the conditions are swapped around, so we first check if it's *not* an evictingPods event. So my suggestion would look somewhat like this:

```go
	var newestEvent *corev1.Event

	for _, event := range eventList.Items {
    if ! IsErrorEvictingPodsEvent(event) {
      continue
    }
		if newestEvent == nil {
			newestEvent = &event
			continue
		}
		if newestEvent.LastTimestamp.Time.Before(event.LastTimestamp.Time) {
			newestEvent = &event
		}
	}

```

Reply:
Done.